### PR TITLE
Add API audit logger

### PR DESCRIPTION
**Pull request by emadkaramad**
This PR adds a new Audit Logger type to allow logging events in DynamoDB using Audit Log API.

### CODE REVIEW

File: src/lib/getAuditLogger.ts [hunk 1]
```diff
@@ -5,6 +5,54 @@ import { isError } from "types/Result"
...
+      } = generateAuditLog(context, action, attributes)
+
+      const userEvent = {
+        eventSource: "User Service",
+        category: "information",
+        eventType: auditAction,
+        timestamp: `${timestamp.toISOString()}`,
+        attributes: {
+          auditLogVersion: "2",
+          eventCode: auditAction.replace(/ /g, "-").toLowerCase(),
```
Comment (daviesjamie) [inline src/lib/getAuditLogger.ts:34]: I think we should carefully think out the event codes, and have an enum or something with them in.

We should come up with a list of all of the audit log events that the user-service will generate, and then manually decide on an appropriate event code for each. I'd expect these manual codes to be a bit clearer than ones that are generated by swapping whitespace for hyphens!

**Review by daviesjamie — CHANGES_REQUESTED** [review]

File: src/lib/getAuditLogger.ts (hunk 1 above)
Comment (emadkaramad) [reply src/lib/getAuditLogger.ts:34]: That makes sense. I'll move the events to an enum.

File: src/types/AuditLogEvent.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    "user-creation.failed-notification",
+    "Error sending notification email of new user creation"
+  )
+
+  public static readonly failedToEmailUserWhileCreatingUser = new AuditLogEvent(
+    "user-creation.failed-emailing-user",
+    "Error sending email to new user"
+  )
+
+  public static readonly editedUser = new AuditLogEvent("user-edit.updated", "Edit user")
```
Comment (daviesjamie) [inline src/types/AuditLogEvent.ts:29]: This is looking much better, but I think we could probably tweak the codes a little to match the format of the ones we already generate in Bichard. Typically they are `<noun>.<verb>` - such as `hearing-outcome.received` or `exceptions.generated`.

Here are some suggestions:

| Event Type | Event Code |
|---|---|
| User logged in | `user.logged-in` |
| Password updated | `user.password.updated` |
| Password reset | `user.password.reset` |
| User deleted | `user.deleted` |
| User created | `user.created` |

I don't know if the "Error creating new user email", "Error sending notification email of new user creation" and "Error sending email to new user" should be audit logs at all - they're more application-level error logs, not something we need to keep for posterity and auditing purposes.

I also think we probably need to look through the user-service and make sure we're audit logging all of the actions users can make - like updating the triggers or details on themselves or other users. But that could be a separate PR/ticket.

**Review by daviesjamie — CHANGES_REQUESTED** [review]

File: src/types/AuditLogEvent.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+  getCode() {
```
Comment (bjpirt) [inline src/types/AuditLogEvent.ts:31]: Do we actually need getters for these? They're public attributes

File: src/types/AuditLogEvent.ts (hunk 2 above)
Comment (emadkaramad) [reply src/types/AuditLogEvent.ts:29]: Thanks for the event codes 🙌. It looks much cleaner and more transparent.
I agree, we probably need to log the failure logs as before in console so logs will be picked by logstash and we can see them in OpenSearch if we need to debug or investigate an issue.

Comment (bjpirt) [reply src/types/AuditLogEvent.ts:29]: 👍 Agree with Jamie's points but looks like you're already on it!

File: src/types/AuditLogEvent.ts (hunk 3 above)
Comment (emadkaramad) [reply src/types/AuditLogEvent.ts:31]: The `code` and `description` fields are private at the moment but I can make them `public readonly` in constructor signature to achieve the same thing and remove `getCode` and `getDescription` functions.
I'll update it.

**Review by daviesjamie — APPROVED** [review]

**Pull request merged**